// piral-cli/lib/common/http.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const request = require("request");
const os_1 = require("os");
const log_1 = require("./log");
const os = os_1.platform();
function getMessage(body) {
    if (body) {
        try {
            const content = JSON.parse(body);
            return content.message;
        }
        catch (ex) {
            return body;
        }
    }
    return '';
}
function postFile(target, key, file) {
    return new Promise(resolve => {
        request(target, {
            method: 'POST',
            headers: {
                authorization: `Basic ${key}`,
                'user-agent': `piral-cli/http.node-${os}`,
            },
        }, (err, res, body) => {
            if (err) {
                log_1.logWarn(err);
                resolve(false);
            }
            else {
                const status = res.statusCode;
                const success = status === 200;
                if (!success) {
                    const message = getMessage(body);
                    log_1.logWarn(`Failed to upload: ${res.statusMessage} (%s).`, status);
                    if (message) {
                        log_1.logWarn(message);
                    }
                }
                resolve(success);
            }
        })
            .form()
            .append('file', file, {
            filename: 'pilet.tgz',
        });
    });
}
exports.postFile = postFile;
