// piral-cli/lib/common/language.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var PiletLanguage;
(function (PiletLanguage) {
    PiletLanguage[PiletLanguage["ts"] = 0] = "ts";
    PiletLanguage[PiletLanguage["js"] = 1] = "js";
})(PiletLanguage = exports.PiletLanguage || (exports.PiletLanguage = {}));
function getLanguageExtension(language) {
    switch (language) {
        case PiletLanguage.js:
            return '.jsx';
        case PiletLanguage.ts:
            return '.tsx';
    }
}
exports.getLanguageExtension = getLanguageExtension;
function getDevDependencies(language) {
    switch (language) {
        case PiletLanguage.ts:
            return {
                typescript: 'latest',
                '@types/react': 'latest',
                '@types/react-dom': 'latest',
                '@types/react-router': 'latest',
                '@types/react-router-dom': 'latest',
                '@types/node': 'latest',
            };
        case PiletLanguage.js:
        default:
            return {};
    }
}
exports.getDevDependencies = getDevDependencies;
