// piral-cli/lib/common/log.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = require("chalk");
function logDebug(message, ...args) {
    console.log(`[debug] ${message}`, ...args.map(arg => chalk_1.default.bold(arg.toString())));
}
exports.logDebug = logDebug;
function logVerbose(message, ...args) {
    console.log(chalk_1.default.gray(message), ...args.map(arg => chalk_1.default.bold(arg.toString())));
}
exports.logVerbose = logVerbose;
function logInfo(message, ...args) {
    console.log(message, ...args.map(arg => chalk_1.default.bold(arg.toString())));
}
exports.logInfo = logInfo;
function logDone(message, ...args) {
    console.log(chalk_1.default.greenBright(message), ...args.map(arg => chalk_1.default.bold(arg.toString())));
}
exports.logDone = logDone;
function logWarn(message, ...args) {
    console.warn(chalk_1.default.red(message), ...args.map(arg => chalk_1.default.bold(arg.toString())));
}
exports.logWarn = logWarn;
function logFail(message, ...args) {
    console.error(chalk_1.default.red(message), ...args.map(arg => chalk_1.default.bold(arg.toString())));
}
exports.logFail = logFail;
