// piral-cli/lib/apps/new-pilet.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const common_1 = require("../common");
exports.newPiletDefaults = {
    target: '.',
    registry: common_1.defaultRegistry,
    source: 'piral',
    forceOverwrite: common_1.ForceOverwrite.no,
    language: common_1.PiletLanguage.ts,
    skipInstall: false,
    template: 'default',
};
function newPilet(baseDir = process.cwd(), options = {}) {
    return __awaiter(this, void 0, void 0, function* () {
        const { target = exports.newPiletDefaults.target, registry = exports.newPiletDefaults.registry, source = exports.newPiletDefaults.source, forceOverwrite = exports.newPiletDefaults.forceOverwrite, language = exports.newPiletDefaults.language, skipInstall = exports.newPiletDefaults.skipInstall, template = exports.newPiletDefaults.template, } = options;
        const root = path_1.resolve(baseDir, target);
        const [sourceName, sourceVersion, hadVersion, type] = yield common_1.dissectPackageName(baseDir, source);
        const success = yield common_1.createDirectory(root);
        if (success) {
            common_1.logInfo(`Scaffolding new pilet in %s ...`, root);
            yield common_1.createFileIfNotExists(root, 'package.json', JSON.stringify({
                name: path_1.basename(root),
                version: '1.0.0',
                description: '',
                keywords: ['pilet'],
                dependencies: {},
                devDependencies: {},
                peerDependencies: {},
                scripts: {},
                main: 'dist/index.js',
                files: ['dist'],
            }, undefined, 2));
            if (registry !== exports.newPiletDefaults.registry) {
                common_1.logInfo(`Setting up NPM registry (%s) ...`, registry);
                yield common_1.createFileIfNotExists(root, '.npmrc', `registry=${registry}
always-auth=true`, forceOverwrite);
            }
            const packageRef = common_1.combinePackageRef(sourceName, sourceVersion, type);
            common_1.logInfo(`Installing NPM package %s ...`, packageRef);
            yield common_1.installPackage(packageRef, root, '--save-dev', '--no-package-lock');
            const packageName = yield common_1.getPackageName(root, sourceName, type);
            const packageVersion = common_1.getPackageVersion(hadVersion, sourceName, sourceVersion, type);
            const piralInfo = yield common_1.readPiralPackage(root, packageName);
            const { piralCLI = { generated: false } } = piralInfo;
            if (!piralCLI.generated) {
                common_1.logWarn(`The used Piral instance does not seem to be a proper development package.
Please make sure to build your development package with the Piral CLI using "piral build".`);
            }
            const { preScaffold, postScaffold } = common_1.getPiletsInfo(piralInfo);
            if (preScaffold) {
                common_1.logInfo(`Running preScaffold script ...`);
                yield common_1.runScript(preScaffold, root);
            }
            common_1.logInfo(`Taking care of templating ...`);
            yield common_1.scaffoldPiletSourceFiles(template, language, root, packageName, forceOverwrite);
            const files = yield common_1.patchPiletPackage(root, packageName, packageVersion, piralInfo, language);
            yield common_1.copyPiralFiles(root, packageName, files, common_1.ForceOverwrite.yes);
            if (!skipInstall) {
                common_1.logInfo(`Installing dependencies ...`);
                yield common_1.installDependencies(root, '--no-package-lock');
            }
            if (postScaffold) {
                common_1.logInfo(`Running postScaffold script ...`);
                yield common_1.runScript(postScaffold, root);
            }
            common_1.logDone(`All done!`);
        }
        else {
            throw new Error('Could not create directory.');
        }
    });
}
exports.newPilet = newPilet;
