// piral-cli/lib/apps/new-piral.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const common_1 = require("../common");
exports.newPiralDefaults = {
    app: './src/index.html',
    onlyCore: false,
    target: '.',
    version: 'latest',
    forceOverwrite: common_1.ForceOverwrite.no,
    language: common_1.PiletLanguage.ts,
    skipInstall: false,
    template: 'default',
};
function newPiral(baseDir = process.cwd(), options = {}) {
    return __awaiter(this, void 0, void 0, function* () {
        const { app = exports.newPiralDefaults.app, onlyCore = exports.newPiralDefaults.onlyCore, target = exports.newPiralDefaults.target, version = exports.newPiralDefaults.version, forceOverwrite = exports.newPiralDefaults.forceOverwrite, language = exports.newPiralDefaults.language, skipInstall = exports.newPiralDefaults.skipInstall, template = exports.newPiralDefaults.template, } = options;
        const root = path_1.resolve(baseDir, target);
        const packageName = onlyCore ? 'piral-core' : 'piral';
        const success = yield common_1.createDirectory(root);
        if (success) {
            const packageRef = common_1.combinePackageRef(packageName, version, 'registry');
            common_1.logInfo(`Creating a new Piral instance in %s ...`, root);
            yield common_1.createFileIfNotExists(root, 'package.json', JSON.stringify({
                name: path_1.basename(root),
                version: '1.0.0',
                description: '',
                keywords: ['piral'],
                dependencies: {},
                scripts: {},
            }, undefined, 2));
            yield common_1.updateExistingJson(root, 'package.json', common_1.getPiralPackage(app, language));
            common_1.logInfo(`Installing NPM package ${packageRef} ...`);
            yield common_1.installPackage(packageRef, root, '--no-package-lock');
            common_1.logInfo(`Taking care of templating ...`);
            yield common_1.scaffoldPiralSourceFiles(template, language, root, app, packageName, forceOverwrite);
            if (!skipInstall) {
                common_1.logInfo(`Installing dependencies ...`);
                yield common_1.installDependencies(root, '--no-package-lock');
            }
            common_1.logDone(`All done!`);
        }
        else {
            throw new Error('Could not create directory.');
        }
    });
}
exports.newPiral = newPiral;
