// piral-cli/lib/commands.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const apps = require("./apps");
exports.apps = apps;
const helpers_1 = require("./helpers");
function specializeCommand(commands, command, suffix) {
    if (command.name.endsWith(suffix)) {
        commands.push(Object.assign(Object.assign({}, command), { name: command.name.replace(suffix, ''), alias: command.alias.filter(n => n.endsWith(suffix)).map(n => n.replace(suffix, '')) }));
    }
}
function specializeCommands(suffix) {
    const commands = [];
    for (const command of allCommands) {
        specializeCommand(commands, command, suffix);
    }
    return commands;
}
const allCommands = [
    {
        name: 'debug-piral',
        alias: ['watch-piral', 'debug-portal', 'watch-portal'],
        description: 'Starts the debugging process for a Piral instance.',
        arguments: ['[source]'],
        // "any" due to https://github.com/microsoft/TypeScript/issues/28663 [artifical N = 50]
        flags(argv) {
            return argv
                .positional('source', {
                type: 'string',
                describe: 'Sets the source root directory or index.html file for collecting all the information.',
                default: apps.debugPiralDefaults.entry,
            })
                .number('port')
                .describe('port', 'Sets the port of the local development server.')
                .default('port', apps.debugPiralDefaults.port)
                .string('cache-dir')
                .describe('cache-dir', 'Sets the cache directory for bundling.')
                .default('cache-dir', apps.debugPiralDefaults.cacheDir)
                .string('public-url')
                .describe('public-url', 'Sets the public URL (path) of the bundle.')
                .default('public-url', apps.debugPiralDefaults.publicUrl)
                .number('log-level')
                .describe('log-level', 'Sets the log level to use (1-5).')
                .default('log-level', apps.debugPiralDefaults.logLevel)
                .boolean('fresh')
                .describe('fresh', 'Resets the cache before starting the debug mode.')
                .default('fresh', apps.debugPiralDefaults.fresh)
                .boolean('open')
                .describe('open', 'Opens the Piral instance directly in the browser.')
                .default('open', apps.debugPiralDefaults.open)
                .boolean('scope-hoist')
                .describe('scope-hoist', 'Tries to reduce bundle size by introducing tree shaking.')
                .default('scope-hoist', apps.debugPiralDefaults.scopeHoist)
                .boolean('hmr')
                .describe('hmr', 'Activates Hot Module Reloading (HMR).')
                .default('hmr', apps.debugPiralDefaults.hmr)
                .boolean('autoinstall')
                .describe('autoinstall', 'Automatically installs missing Node.js packages.')
                .default('autoinstall', apps.debugPiralDefaults.autoInstall)
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.debugPiral(args.base, {
                entry: args.source,
                cacheDir: args.cacheDir,
                port: args.port,
                hmr: args.hmr,
                autoInstall: args.autoinstall,
                scopeHoist: args.scopeHoist,
                publicUrl: args.publicUrl,
                logLevel: args.logLevel,
                fresh: args.fresh,
                open: args.open,
            });
        },
    },
    {
        name: 'build-piral',
        alias: ['bundle-piral', 'build-portal', 'bundle-portal'],
        description: 'Creates a production build for a Piral instance.',
        arguments: ['[source]'],
        // "any" due to https://github.com/microsoft/TypeScript/issues/28663 [artifical N = 50]
        flags(argv) {
            return argv
                .positional('source', {
                type: 'string',
                describe: 'Sets the source root directory or index.html file for collecting all the information.',
                default: apps.buildPiralDefaults.entry,
            })
                .string('target')
                .describe('target', 'Sets the target directory or file of bundling.')
                .default('target', apps.buildPiralDefaults.target)
                .string('cache-dir')
                .describe('cache-dir', 'Sets the cache directory for bundling.')
                .default('cache-dir', apps.buildPiralDefaults.cacheDir)
                .string('public-url')
                .describe('public-url', 'Sets the public URL (path) of the bundle.')
                .default('public-url', apps.buildPiralDefaults.publicUrl)
                .boolean('detailed-report')
                .describe('detailed-report', 'Sets if a detailed report should be created.')
                .default('detailed-report', apps.buildPiralDefaults.detailedReport)
                .number('log-level')
                .describe('log-level', 'Sets the log level to use (1-5).')
                .default('log-level', apps.buildPiralDefaults.logLevel)
                .boolean('fresh')
                .describe('fresh', 'Performs a fresh build by removing the target directory first.')
                .default('fresh', apps.buildPiralDefaults.fresh)
                .boolean('minify')
                .describe('minify', 'Performs minification or other post-bundle transformations.')
                .default('minify', apps.buildPiralDefaults.minify)
                .boolean('source-maps')
                .describe('source-maps', 'Create associated source maps for the bundles.')
                .default('source-maps', apps.buildPiralDefaults.sourceMaps)
                .boolean('content-hash')
                .describe('content-hash', 'Appends the hash to the side-bundle files.')
                .default('content-hash', apps.buildPiralDefaults.contentHash)
                .boolean('scope-hoist')
                .describe('scope-hoist', 'Tries to reduce bundle size by introducing tree shaking.')
                .default('scope-hoist', apps.buildPiralDefaults.scopeHoist)
                .choices('type', ['all', 'release', 'develop'])
                .describe('type', 'Selects the target type of the build. "all" builds all target types.')
                .default('type', apps.buildPiralDefaults.type)
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.buildPiral(args.base, {
                entry: args.source,
                target: args.target,
                cacheDir: args.cacheDir,
                publicUrl: args.publicUrl,
                minify: args.minify,
                scopeHoist: args.scopeHoist,
                contentHash: args.contentHash,
                sourceMaps: args.sourceMaps,
                detailedReport: args.detailedReport,
                logLevel: args.logLevel,
                type: args.type,
            });
        },
    },
    {
        name: 'new-piral',
        alias: ['create-piral', 'scaffold-piral', 'setup-piral'],
        description: 'Creates a new Piral instance by adding all files and changes to the current project.',
        arguments: ['[target]'],
        flags(argv) {
            return argv
                .positional('target', {
                type: 'string',
                describe: "Sets the project's root directory for making the changes.",
                default: apps.newPiralDefaults.target,
            })
                .string('app')
                .describe('app', "Sets the path to the app's source HTML file.")
                .default('app', apps.newPiralDefaults.app)
                .boolean('only-core')
                .describe('only-core', 'Sets if "piral-core" should be used. Otherwise, "piral" is used.')
                .default('only-core', apps.newPiralDefaults.onlyCore)
                .boolean('skip-install')
                .describe('skip-install', 'Skips the installation of the dependencies using NPM.')
                .default('skip-install', apps.newPiralDefaults.skipInstall)
                .string('tag')
                .describe('tag', 'Sets the tag or version of the package to install. By default, it is "latest".')
                .default('tag', apps.newPiralDefaults.version)
                .choices('force-overwrite', helpers_1.forceOverwriteKeys)
                .describe('force-overwrite', 'Determines if files should be overwritten by the installation.')
                .default('force-overwrite', helpers_1.keyOfForceOverwrite(apps.newPiralDefaults.forceOverwrite))
                .choices('language', helpers_1.piletLanguageKeys)
                .describe('language', 'Determines the programming language for the new Piral instance.')
                .default('language', helpers_1.keyOfPiletLanguage(apps.newPiralDefaults.language))
                .choices('template', helpers_1.templateTypeKeys)
                .describe('template', 'Sets the boilerplate template to be used when scaffolding.')
                .default('template', helpers_1.templateTypeKeys[0])
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.newPiral(args.base, {
                app: args.app,
                target: args.target,
                onlyCore: args.onlyCore,
                version: args.tag,
                forceOverwrite: helpers_1.valueOfForceOverwrite(args.forceOverwrite),
                language: helpers_1.valueOfPiletLanguage(args.language),
                skipInstall: args.skipInstall,
                template: args.template,
            });
        },
    },
    {
        name: 'validate-piral',
        alias: ['verify-piral', 'check-piral'],
        description: 'Checks the validity of the current project as a Piral instance.',
        arguments: ['[source]'],
        flags(argv) {
            return argv
                .positional('source', {
                type: 'string',
                describe: 'Sets the source root directory or index.html file for collecting all the information.',
                default: apps.validatePiralDefaults.entry,
            })
                .number('log-level')
                .describe('log-level', 'Sets the log level to use (1-5).')
                .default('log-level', apps.validatePiralDefaults.logLevel)
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.validatePiral(args.base, {
                entry: args.entry,
                logLevel: args.logLevel,
            });
        },
    },
    {
        name: 'debug-pilet',
        alias: ['watch-pilet', 'debug', 'watch'],
        description: 'Starts the debugging process for a pilet using a Piral instance.',
        arguments: ['[source]'],
        // "any" due to https://github.com/microsoft/TypeScript/issues/28663 [artifical N = 50]
        flags(argv) {
            return argv
                .positional('source', {
                type: 'string',
                describe: 'Sets the source file containing the pilet root module.',
                default: apps.debugPiletDefaults.entry,
            })
                .number('port')
                .describe('port', 'Sets the port of the local development server.')
                .default('port', apps.debugPiletDefaults.port)
                .string('cache-dir')
                .describe('cache-dir', 'Sets the cache directory for bundling.')
                .default('cache-dir', apps.debugPiletDefaults.cacheDir)
                .number('log-level')
                .describe('log-level', 'Sets the log level to use (1-5).')
                .default('log-level', apps.debugPiletDefaults.logLevel)
                .boolean('fresh')
                .describe('fresh', 'Resets the cache before starting the debug mode.')
                .default('fresh', apps.debugPiletDefaults.fresh)
                .boolean('open')
                .describe('open', 'Opens the pilet directly in the browser.')
                .default('open', apps.debugPiletDefaults.open)
                .boolean('scope-hoist')
                .describe('scope-hoist', 'Tries to reduce bundle size by introducing tree shaking.')
                .default('scope-hoist', apps.debugPiletDefaults.scopeHoist)
                .boolean('hmr')
                .describe('hmr', 'Activates Hot Module Reloading (HMR).')
                .default('hmr', apps.debugPiletDefaults.hmr)
                .boolean('autoinstall')
                .describe('autoinstall', 'Automatically installs missing Node.js packages.')
                .default('autoinstall', apps.debugPiletDefaults.autoInstall)
                .string('app')
                .describe('app', 'Sets the name of the Piral instance.')
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.debugPilet(args.base, {
                entry: args.source,
                cacheDir: args.cacheDir,
                port: args.port,
                scopeHoist: args.scopeHoist,
                hmr: args.hmr,
                autoInstall: args.autoinstall,
                app: args.app,
                logLevel: args.logLevel,
                fresh: args.fresh,
                open: args.open,
            });
        },
    },
    {
        name: 'build-pilet',
        alias: ['bundle-pilet', 'build', 'bundle'],
        description: 'Creates a production build for a pilet.',
        arguments: ['[source]'],
        // "any" due to https://github.com/microsoft/TypeScript/issues/28663 [artifical N = 50]
        flags(argv) {
            return argv
                .positional('source', {
                type: 'string',
                describe: 'Sets the source index.tsx file for collecting all the information.',
                default: apps.buildPiletDefaults.entry,
            })
                .string('target')
                .describe('target', 'Sets the target file of bundling.')
                .default('target', apps.buildPiletDefaults.target)
                .string('cache-dir')
                .describe('cache-dir', 'Sets the cache directory for bundling.')
                .default('cache-dir', apps.buildPiletDefaults.cacheDir)
                .boolean('detailed-report')
                .describe('detailed-report', 'Sets if a detailed report should be created.')
                .default('detailed-report', apps.buildPiletDefaults.detailedReport)
                .number('log-level')
                .describe('log-level', 'Sets the log level to use (1-5).')
                .default('log-level', apps.buildPiletDefaults.logLevel)
                .boolean('fresh')
                .describe('fresh', 'Performs a fresh build by removing the target directory first.')
                .default('fresh', apps.buildPiletDefaults.fresh)
                .boolean('minify')
                .describe('minify', 'Performs minification or other post-bundle transformations.')
                .default('minify', apps.buildPiletDefaults.minify)
                .boolean('source-maps')
                .describe('source-maps', 'Creates source maps for the bundles.')
                .default('source-maps', apps.buildPiletDefaults.sourceMaps)
                .boolean('content-hash')
                .describe('content-hash', 'Appends the hash to the side-bundle files.')
                .default('content-hash', apps.buildPiletDefaults.contentHash)
                .boolean('scope-hoist')
                .describe('scope-hoist', 'Tries to reduce bundle size by introducing tree shaking.')
                .default('scope-hoist', apps.buildPiletDefaults.scopeHoist)
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.buildPilet(args.base, {
                entry: args.source,
                target: args.target,
                cacheDir: args.cacheDir,
                minify: args.minify,
                contentHash: args.contentHash,
                sourceMaps: args.sourceMaps,
                scopeHoist: args.scopeHoist,
                detailedReport: args.detailedReport,
                fresh: args.fresh,
                logLevel: args.logLevel,
            });
        },
    },
    {
        name: 'pack-pilet',
        alias: ['package-pilet', 'pack', 'package'],
        description: 'Creates a pilet package that can be published.',
        arguments: ['[source]'],
        flags(argv) {
            return argv
                .positional('source', {
                type: 'string',
                describe: 'Sets the source package.json file for creating the package.',
                default: apps.packPiletDefaults.source,
            })
                .string('target')
                .describe('target', 'Sets the target directory or file of packing.')
                .default('target', apps.packPiletDefaults.target)
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.packPilet(args.base, {
                source: args.source,
                target: args.target,
            });
        },
    },
    {
        name: 'publish-pilet',
        alias: ['post-pilet', 'publish'],
        description: 'Publishes a pilet package to a pilet feed.',
        arguments: ['[source]'],
        flags(argv) {
            return argv
                .positional('source', {
                type: 'string',
                describe: 'Sets the source previously packed *.tgz bundle to publish.',
                default: apps.publishPiletDefaults.source,
            })
                .string('url')
                .describe('url', 'Sets the explicit URL where to publish the pilet to.')
                .default('url', apps.publishPiletDefaults.url)
                .string('api-key')
                .describe('api-key', 'Sets the potential API key to send to the service.')
                .default('api-key', apps.publishPiletDefaults.apiKey)
                .boolean('fresh')
                .describe('fresh', 'Performs a fresh build, then packages and finally publishes the pilet.')
                .default('fresh', apps.publishPiletDefaults.fresh)
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.')
                .demandOption('url');
        },
        run(args) {
            return apps.publishPilet(args.base, {
                source: args.source,
                apiKey: args.apiKey,
                url: args.url,
                fresh: args.fresh,
            });
        },
    },
    {
        name: 'new-pilet',
        alias: ['create-pilet', 'scaffold-pilet', 'scaffold', 'new', 'create'],
        description: 'Scaffolds a new pilet for a specified Piral instance.',
        arguments: ['[source]'],
        flags(argv) {
            return argv
                .positional('source', {
                type: 'string',
                describe: 'Sets the source package containing a Piral instance for templating the scaffold process.',
                default: apps.newPiletDefaults.source,
            })
                .string('target')
                .describe('target', 'Sets the target directory for scaffolding. By default, the current directory.')
                .default('target', apps.newPiletDefaults.target)
                .string('registry')
                .describe('registry', 'Sets the package registry to use for resolving the specified Piral app.')
                .default('registry', apps.newPiletDefaults.registry)
                .boolean('skip-install')
                .describe('skip-install', 'Skips the installation of the dependencies using NPM.')
                .default('skip-install', apps.newPiletDefaults.skipInstall)
                .choices('force-overwrite', helpers_1.forceOverwriteKeys)
                .describe('force-overwrite', 'Determines if files should be overwritten by the scaffolding.')
                .default('force-overwrite', helpers_1.keyOfForceOverwrite(apps.newPiletDefaults.forceOverwrite))
                .choices('language', helpers_1.piletLanguageKeys)
                .describe('language', 'Determines the programming language for the new pilet.')
                .default('language', helpers_1.keyOfPiletLanguage(apps.newPiletDefaults.language))
                .choices('template', helpers_1.templateTypeKeys)
                .describe('template', 'Sets the boilerplate template to be used when scaffolding.')
                .default('template', helpers_1.templateTypeKeys[0])
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.newPilet(args.base, {
                target: args.target,
                source: args.source,
                registry: args.registry,
                forceOverwrite: helpers_1.valueOfForceOverwrite(args.forceOverwrite),
                language: helpers_1.valueOfPiletLanguage(args.language),
                skipInstall: args.skipInstall,
                template: args.template,
            });
        },
    },
    {
        name: 'upgrade-pilet',
        alias: ['upgrade'],
        description: 'Upgrades an existing pilet to the latest version of the used Piral instance.',
        arguments: [],
        flags(argv) {
            return argv
                .string('target')
                .describe('target', 'Sets the target directory to upgrade. By default, the current directory.')
                .default('target', apps.upgradePiletDefaults.target)
                .string('tag')
                .describe('tag', 'Sets the tag or version of the Piral instance to upgrade to. By default, it is "latest".')
                .default('tag', apps.upgradePiletDefaults.version)
                .choices('force-overwrite', helpers_1.forceOverwriteKeys)
                .describe('force-overwrite', 'Determines if files should be overwritten by the upgrading process.')
                .default('force-overwrite', helpers_1.keyOfForceOverwrite(apps.upgradePiletDefaults.forceOverwrite))
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.upgradePilet(args.base, {
                target: args.target,
                version: args.tag,
                forceOverwrite: helpers_1.valueOfForceOverwrite(args.forceOverwrite),
            });
        },
    },
    {
        name: 'validate-pilet',
        alias: ['verify-pilet', 'check-pilet', 'lint-pilet', 'assert-pilet'],
        description: 'Checks the validity of the current pilet according to the rules defined by the Piral instance.',
        arguments: ['[source]'],
        flags(argv) {
            return argv
                .positional('source', {
                type: 'string',
                describe: 'Sets the source file containing the pilet root module.',
                default: apps.validatePiletDefaults.entry,
            })
                .number('log-level')
                .describe('log-level', 'Sets the log level to use (1-5).')
                .default('log-level', apps.validatePiletDefaults.logLevel)
                .string('app')
                .describe('app', 'Sets the name of the Piral instance.')
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.validatePilet(args.base, {
                entry: args.entry,
                logLevel: args.logLevel,
                app: args.app,
            });
        },
    },
];
class Commands {
    constructor() {
        this.all = allCommands;
    }
    get pilet() {
        return specializeCommands('-pilet');
    }
    get piral() {
        return specializeCommands('-piral');
    }
}
exports.commands = new Commands();
